homeassistant/components/ws66i/__init__.py

```diff
@@ -0,0 +1,127 @@
+"""The Soundavo WS66i 6-Zone Amplifier integration."""
+from __future__ import annotations
+
+import logging
+
+from pyws66i import WS66i, get_ws66i
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_IP_ADDRESS, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from .const import CONF_SOURCES, DOMAIN, INIT_OPTIONS_DEFAULT
+from .coordinator import Ws66iDataUpdateCoordinator
+from .models import SourceRep, Ws66iData
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = ["media_player"]
+
+
+@callback
+def _get_sources_from_dict(data) -> SourceRep:
+    sources_config = data[CONF_SOURCES]
+
+    # Dict index to custom name
+    source_id_name = {int(index): name for index, name in sources_config.items()}
+
+    # Dict custom name to index
+    source_name_id = {v: k for k, v in source_id_name.items()}
+
+    # List of custom names
+    source_names = sorted(source_name_id.keys(), key=lambda v: source_name_id[v])
+
+    return SourceRep(source_id_name, source_name_id, source_names)
+
+
+def _find_zones(hass: HomeAssistant, ws66i: WS66i) -> list[int]:
+    """Generate zones list by searching for presence of zones."""
+    # Zones 11 - 16 are the master amp
+    # Zones 21,31 - 26,36 are the daisy-chained amps
+    zone_list = []
+    for amp_num in range(1, 4):
+
+        if amp_num > 1:
+            # Don't add entities that aren't present
+            status = ws66i.zone_status(amp_num * 10 + 1)
+            if status is None:
+                break
+
+        for zone_num in range(1, 7):
+            zone_id = (amp_num * 10) + zone_num
+            zone_list.append(zone_id)
+
+    _LOGGER.info("Detected %d amp(s)", amp_num - 1)
+    return zone_list
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Soundavo WS66i 6-Zone Amplifier from a config entry."""
+    # Check if options was set. If not, initialize a dict for it
+    options: dict[str, dict[str, str]]
+    if not entry.options:
+        options = {CONF_SOURCES: INIT_OPTIONS_DEFAULT}
+        hass.config_entries.async_update_entry(entry, options=options)
+    else:
```

Comment:
You can avoid this complexity by setting `options=` when you do `async_create_entry` in the config flow